### PR TITLE
feat(tgw): added ec2 tgw & tgw-vpc-attachment

### LINKED ISSUE #363: Support for AWS TransitGateway and TransitGatewayAttachment
### What problem are you facing?

in my team we spin up clusters per default with internal load balancers + we have multiple clusters in different AWS accounts.
to make them available to the corp intranet + to VPN users etc. we are doing some transit gateway peering afterwards to make a cluster: 

- availability  to the internal network/routing 
- ensure visibility  to other parts of the infrastructure(CI/CID) in different AWS accounts 
- enable the new cluster to also reach other pieces of infra. like CI/CD, services in other clusters or RDS instances 
- also: we create a 2nd VPC in the same account which contains only RDS instances for better separation of concerns(well, that may not matter)

currently we need to run dedicated pipeline tasks after the cluster creation to apply _terraform code_ which is doing the necessary TGW peering.  
we are currently using SAP gardener as our cluster lifecycle management tool. gardener provides CRD's to CRUD the clusters. 

_argoCD_ is our tool to apply those manifests against a management(seed) cluster.
_argoCD_ is also our tool of choice for applying additional landscape deployments like prometheus-operator, nginx etc. 
now if we want to do the TGW peering we need to go one step back and involve some tooling(tekton pipelines) to apply the terraform code against the AWS account. 

### How could Crossplane help solve your problem?

would be super cool to also deploy those infrastructure changes/resources via _argoCD_ and kubernetes manifests. 

see: https://crossplane.slack.com/archives/C01718T2476/p1601571788009400
and: https://kubernetes.slack.com/archives/CD6U2V71N/p1601564676059800

i think you get the idea. :) 

thanks in ahead for considering this feature request.

### PR DESCRIPTION
**Pull request by haarchri**
Signed-off-by: Christopher Haar <chhaar30@googlemail.com>
### Description of your changes

- implements ec2 tgw (transitgateway)
- implements ec2 tgw-vpc-attachment (transitgateway-vpc-attachment)

**transitgateway:**

```
 kubectl describe transitgateway.ec2.aws.crossplane.io/tgw  
Name:         tgw
Namespace:    
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2021-12-13T21:52:07+01:00
              crossplane.io/external-create-succeeded: 2021-12-13T21:52:07+01:00
              crossplane.io/external-name: tgw-0cf9ff0c1a0e88782
API Version:  ec2.aws.crossplane.io/v1alpha1
Kind:         TransitGateway
Metadata:
  Creation Timestamp:  2021-12-13T20:52:06Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  3
  Managed Fields:
    API Version:  ec2.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:deletionPolicy:
        f:forProvider:
          .:
          f:region:
        f:providerConfigRef:
          .:
          f:name:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-12-13T20:52:06Z
    API Version:  ec2.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-create-pending:
          f:crossplane.io/external-create-succeeded:
          f:crossplane.io/external-name:
        f:finalizers:
          .:
          v:"finalizer.managedresource.crossplane.io":
      f:spec:
        f:forProvider:
          f:options:
            .:
            f:amazonSideASN:
            f:autoAcceptSharedAttachments:
            f:defaultRouteTableAssociation:
            f:defaultRouteTablePropagation:
            f:dnsSupport:
            f:multicastSupport:
            f:vpnECMPSupport:
          f:tags:
      f:status:
        .:
        f:atProvider:
          .:
          f:creationTime:
          f:ownerID:
          f:state:
          f:transitGatewayARN:
          f:transitGatewayID:
        f:conditions:
    Manager:         provider
    Operation:       Update
    Time:            2021-12-13T20:52:07Z
  Resource Version:  1607723
  UID:               a7f67c80-87be-4e5f-a2c9-ba8e83cf52e7
Spec:
  Deletion Policy:  Delete
  For Provider:
    Options:
      Amazon Side ASN:                  64512
      Auto Accept Shared Attachments:   disable
      Default Route Table Association:  enable
      Default Route Table Propagation:  enable
      Dns Support:                      enable
      Multicast Support:                disable
      Vpn ECMP Support:                 enable
    Region:                             us-east-1
    Tags:
      Key:    crossplane-kind
      Value:  transitgateway.ec2.aws.crossplane.io
      Key:    crossplane-name
      Value:  tgw
      Key:    crossplane-providerconfig
      Value:  example
  Provider Config Ref:
    Name:  example
Status:
  At Provider:
    Creation Time:        2021-12-13T20:52:07Z
    Owner ID:             255932642927
    State:                available
    Transit Gateway ARN:  arn:aws:ec2:us-east-1:255932642927:transit-gateway/tgw-0cf9ff0c1a0e88782
    Transit Gateway ID:   tgw-0cf9ff0c1a0e88782
  Conditions:
    Last Transition Time:  2021-12-13T20:54:09Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-12-13T20:52:07Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
```

**transitgateway-vpc-attachment:**

```
kubectl describe transitgatewayvpcattachment.ec2.aws.crossplane.io/tgw-vpc-attach 
Name:         tgw-vpc-attach
Namespace:    
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2021-12-13T21:55:06+01:00
              crossplane.io/external-create-succeeded: 2021-12-13T21:55:08+01:00
              crossplane.io/external-name: tgw-attach-03fb604eaeeaef23e
API Version:  ec2.aws.crossplane.io/v1alpha1
Kind:         TransitGatewayVPCAttachment
Metadata:
  Creation Timestamp:  2021-12-13T20:55:06Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  3
  Managed Fields:
    API Version:  ec2.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:deletionPolicy:
        f:forProvider:
          .:
          f:region:
          f:subnetIdsRefs:
          f:transitGatewayIdRef:
            .:
            f:name:
          f:vpcIdRef:
            .:
            f:name:
        f:providerConfigRef:
          .:
          f:name:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-12-13T20:55:06Z
    API Version:  ec2.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-create-pending:
          f:crossplane.io/external-create-succeeded:
          f:crossplane.io/external-name:
        f:finalizers:
          .:
          v:"finalizer.managedresource.crossplane.io":
      f:spec:
        f:forProvider:
          f:subnetIds:
          f:tags:
          f:transitGatewayId:
          f:vpcId:
      f:status:
        .:
        f:atProvider:
          .:
          f:creationTime:
          f:state:
          f:subnetIDs:
          f:transitGatewayAttachmentID:
          f:transitGatewayID:
          f:vpcID:
          f:vpcOwnerID:
        f:conditions:
    Manager:         provider
    Operation:       Update
    Time:            2021-12-13T20:55:08Z
  Resource Version:  1608221
  UID:               1f8aef5a-2676-4215-9fc8-8f6d94809426
Spec:
  Deletion Policy:  Delete
  For Provider:
    Region:  us-east-1
    Subnet Ids:
      subnet-0bea8bdb2151f4fbe
      subnet-04be0211e759fb076
    Subnet Ids Refs:
      Name:  sample-subnet1
      Name:  sample-subnet2
    Tags:
      Key:               crossplane-kind
      Value:             transitgatewayvpcattachment.ec2.aws.crossplane.io
      Key:               crossplane-name
      Value:             tgw-vpc-attach
      Key:               crossplane-providerconfig
      Value:             example
    Transit Gateway Id:  tgw-0cf9ff0c1a0e88782
    Transit Gateway Id Ref:
      Name:  tgw
    Vpc Id:  vpc-0557e8b3073ddddad
    Vpc Id Ref:
      Name:  sample-vpc
  Provider Config Ref:
    Name:  example
Status:
  At Provider:
    Creation Time:  2021-12-13T20:55:08Z
    State:          available
    Subnet I Ds:
      subnet-0bea8bdb2151f4fbe
      subnet-04be0211e759fb076
    Transit Gateway Attachment ID:  tgw-attach-03fb604eaeeaef23e
    Transit Gateway ID:             tgw-0cf9ff0c1a0e88782
    Vpc ID:                         vpc-0557e8b3073ddddad
    Vpc Owner ID:                   255932642927
  Conditions:
    Last Transition Time:  2021-12-13T20:57:09Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-12-13T20:55:08Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced

```

Fixes #363 #750

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by AaronME** [comment]
@haarchri Is this still WIP or ready for review?

**Comment by haarchri** [comment]
@AaronME  ready for review - thanks for the hint in the topic description

**Comment by haarchri** [comment]
still blocked by #920

**Comment by haarchri** [comment]
rebased & tested again after #920  was merged


```
NAME                                       READY   SYNCED   EXTERNAL-NAME
transitgateway.ec2.aws.crossplane.io/tgw   True    True     tgw-0d0013ce6d364d20b

NAME                                                               READY   SYNCED   EXTERNAL-NAME
transitgatewayvpcattachment.ec2.aws.crossplane.io/tgw-vpc-attach   True    True     tgw-attach-0bc3d1f136a3368a3
```

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:42]:
```suggestion
	// +optional
	// +crossplane:generate:reference:type=VPC
```

Can we use generated resolvers for these referenced fields?

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:43]: Nitpick: while we don't have explicit convention around this, I think it's fair to go with calculated lower camel, i.e. `vpcId`, instead of taking known words into account. There are some places we do one and other places we do the other, but I believe calculated lower camel is the most consistent for repos where there is code generation, like this one or Jet-based providers.

File: pkg/controller/ec2/transitgateway/setup.go
Comment (muvaf) [inline pkg/controller/ec2/transitgateway/setup.go:100]: Let's add a [`tagger`](https://github.com/crossplane/provider-aws/blob/master/pkg/controller/ecr/repository/controller.go#L242) to do this during initialization.

File: pkg/controller/ec2/transitgateway/setup.go
Comment (muvaf) [inline pkg/controller/ec2/transitgateway/setup.go:131]: We should make sure the CR does not have a value for given field before we late initialize. It seems we assign whatever we get from AWS, which could mean overriding user parameters.

File: pkg/controller/ec2/transitgatewayvpcattachment/setup.go
Comment (muvaf) [inline pkg/controller/ec2/transitgatewayvpcattachment/setup.go:88]: Similarly, a `tagger` can do that.

File: pkg/controller/ec2/transitgatewayvpcattachment/setup.go
Comment (muvaf) [inline pkg/controller/ec2/transitgatewayvpcattachment/setup.go:123]:
```suggestion
```

Its default is no-op already.

**Review by muvaf — COMMENTED** [review]
I have a few comments but I think we'll get to merge it before the release.

**Comment by haarchri** [comment]
@muvaf  changed your suggestions

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:109]:
```suggestion
```
nitpick: these are defaults already

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:140]:
```suggestion
	// +crossplane:generate:reference:refFieldName=SubnetIDRefs
	// +crossplane:generate:reference:selectorFieldName=SubnetIDSelector
	SubnetIDs []*string `json:"subnetIds,omitempty"`

	// SubnetIDsRefs is a list of references to SubnetIDs used to set
	// the SubnetIDs.
	// +optional
	SubnetIDRefs []xpv1.Reference `json:"subnetIdRefs,omitempty"`

	// SubnetIDSelector selects references to SubnetIDs used
	// to set the SubnetIDs.
	// +optional
	SubnetIDSelector *xpv1.Selector `json:"subnetIdSelector,omitempty"`
```

We usually go for plural in `Refs` instead of the field in case of arrays.

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:146]:
```suggestion
```

nitpick

File: examples/ec2/transitgatewayvpcattachment.yaml
Comment (muvaf) [inline examples/ec2/transitgatewayvpcattachment.yaml:8]:
```suggestion
    subnetIdRefs:
```

File: pkg/controller/ec2/transitgatewayvpcattachment/setup.go
Comment (muvaf) [inline pkg/controller/ec2/transitgatewayvpcattachment/setup.go:60]:
```suggestion
	resp := &svcsdk.DescribeTransitGatewayVpcAttachmentsOutput{}
	for _, transitGatewayVpcAttachment := range obj.TransitGatewayVpcAttachments {
		if aws.StringValue(transitGatewayVpcAttachment.TransitGatewayAttachmentId) == meta.GetExternalName(cr) {
```

File: pkg/controller/ec2/vpcpeeringconnection/setup.go
Comment (muvaf) [inline pkg/controller/ec2/vpcpeeringconnection/setup.go:55]:
```suggestion
			managed.WithInitializers(managed.NewNameAsExternalName(mgr.GetClient()), &tagger{kube: mgr.GetClient()}),
```

Provider config default is deprecated in favor of CRD defaulting.

File: apis/ec2/v1alpha1/custom_types.go
Comment (muvaf) [inline apis/ec2/v1alpha1/custom_types.go:70]: These are breaking changes so we'd need to bump `VPCPeeringConnection` to `v1alpha2` or `v1beta1`

**Review by muvaf — APPROVED** [review]
I have a couple of comments but LGTM in general. As synced, having `VPCPeeringConnection` changes in a separate PR would be better.

**Comment by haarchri** [comment]
i will open fullow up PR for cleanup vpc_peering_connection - thanks for review

**Pull request merged**
